examples/mnist/Main.hs

```diff
@@ -0,0 +1,159 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE UndecidableSuperClasses #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE NoStarIsType #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE OverloadedLists #-}
+
+module Main where
+
+import           Control.Monad                  ( foldM
+                                                , when
+                                                )
+import           Control.Exception.Safe         (try, SomeException(..))
+import           GHC.Generics
+import           GHC.TypeLits
+
+import           Torch.Static
+import           Torch.Static.Native     hiding ( linear )
+import           Torch.Static.Factories
+import qualified Torch.Autograd                as A
+import qualified Torch.NN                      as A
+import qualified Torch.DType                   as D
+import qualified Torch.Tensor                  as D
+import qualified Torch.Functions               as D
+import qualified Torch.TensorFactories         as D
+import qualified Image                         as I
+import           System.Random
+import           System.Environment
+
+--------------------------------------------------------------------------------
+-- Multi-Layer Perceptron (MLP)
+--------------------------------------------------------------------------------
+
+
+newtype Parameter dtype shape = Parameter A.IndependentTensor deriving (Show)
+
+toDependent :: Parameter dtype shape -> Tensor dtype shape
+toDependent (Parameter t) = UnsafeMkTensor $ A.toDependent t
+
+instance A.Parameterized (Parameter dtype shape) where
+  flattenParameters (Parameter x) = [x]
+  replaceOwnParameters _ = Parameter <$> A.nextParameter
+
+data LinearSpec (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) = LinearSpec
+  deriving (Show, Eq)
+
+data Linear (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) =
+  Linear { weight :: Parameter dtype '[inputFeatures, outputFeatures]
+         , bias :: Parameter dtype '[outputFeatures]
+         } deriving (Show, Generic)
+
+linear
+  :: Linear dtype inputFeatures outputFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+linear Linear {..} input =
+  add (mm input (toDependent weight)) (toDependent bias)
+
+makeIndependent :: Tensor dtype shape -> IO (Parameter dtype shape)
+makeIndependent t = Parameter <$> A.makeIndependent (toDynamic t)
+
+instance A.Parameterized (Linear dtype inputFeatures outputFeatures)
+
+instance (KnownDType dtype, KnownNat inputFeatures, KnownNat outputFeatures) => A.Randomizable (LinearSpec dtype inputFeatures outputFeatures) (Linear dtype inputFeatures outputFeatures) where
+  sample LinearSpec =
+    Linear <$> (makeIndependent =<< randn) <*> (makeIndependent =<< randn)
+
```

Comment:
Lines 47-79 seem like the should go into the static version of NN in the library instead of being here, though I'd be fine with merging first and saving the library merge  for a subsequent PR.

Reply:
let's postpone that to later. these lines appear in a number of examples. We will need to revisit this again for type-safe gradients, anyway.